proposals/infra/INF-0002-root-signature-in-clang.md

```diff
@@ -0,0 +1,683 @@
+  MAXIMUM_MIN_LINEAR_MAG_POINT_MIP_LINEAR = 0x191,
+  MAXIMUM_MIN_MAG_LINEAR_MIP_POINT = 0x194,
+  MAXIMUM_MIN_MAG_MIP_LINEAR = 0x195,
+  MAXIMUM_ANISOTROPIC = 0x1d5
+};
+enum class ShaderVisibility : uint32_t {
+  All = 0,
+  Vertex = 1,
+  Hull = 2,
+  Domain = 3,
+  Geometry = 4,
+  Pixel = 5,
+  Amplification = 6,
+  Mesh = 7,
+  MaxValue = 7
+};
+enum class StaticBorderColor : uint32_t {
+  TransparentBlack = 0,
+  OpaqueBlack = 1,
+  OpaqueWhite = 2,
+  OpaqueBlackUint = 3,
+  OpaqueWhiteUint = 4
+};
+enum class TextureAddressMode : uint32_t {
+  Wrap = 1,
+  Mirror = 2,
+  Clamp = 3,
+  Border = 4,
+  MirrorOnce = 5
+};
+
+struct ContainerRootDescriptor {
+  uint32_t ShaderRegister;
+  uint32_t RegisterSpace = 0;
+  uint32_t Flags;
+};
+
+struct ContainerDescriptorRange {
+  hlsl::RootSignature::DescriptorRangeType RangeType;
+  uint32_t NumDescriptors;
+  uint32_t BaseShaderRegister;
+  uint32_t RegisterSpace;
+  uint32_t Flags;
+  uint32_t OffsetInDescriptorsFromTableStart;
+};
+
+struct ContainerRootDescriptorTable {
+  uint32_t NumDescriptorRanges;
+};
+
+struct ContainerRootConstants {
+  uint32_t ShaderRegister;
+  uint32_t RegisterSpace = 0;
+  uint32_t Num32BitValues;
+};
+
+struct ContainerRootParameter {
+  hlsl::RootSignature::RootParameterType ParameterType;
+  hlsl::RootSignature::ShaderVisibility ShaderVisibility =
+      hlsl::RootSignature::ShaderVisibility::All;
+};
+
+struct StaticSamplerDesc {
+  hlsl::RootSignature::Filter Filter =
+      hlsl::RootSignature::Filter::ANISOTROPIC;
+  hlsl::RootSignature::TextureAddressMode AddressU =
+      hlsl::RootSignature::TextureAddressMode::Wrap;
+  hlsl::RootSignature::TextureAddressMode AddressV =
+      hlsl::RootSignature::TextureAddressMode::Wrap;
+  hlsl::RootSignature::TextureAddressMode AddressW =
+      hlsl::RootSignature::TextureAddressMode::Wrap;
+  float MipLODBias = 0;
+  uint32_t MaxAnisotropy = 16;
+  hlsl::RootSignature::ComparisonFunc ComparisonFunc =
+      hlsl::RootSignature::ComparisonFunc::LessEqual;
+  hlsl::RootSignature::StaticBorderColor BorderColor =
+      hlsl::RootSignature::StaticBorderColor::OpaqueWhite;
+  float MinLOD = 0;
+  float MaxLOD = hlsl::RootSignature::Float32Max;
+  uint32_t ShaderRegister;
+  uint32_t RegisterSpace = 0;
+  hlsl::RootSignature::ShaderVisibility ShaderVisibility =
+      hlsl::RootSignature::ShaderVisibility::All;
+};
+
+struct ParsedRootSignature {
+  dxbc::RootSignature::ContainerRootSignatureDesc RSDesc;
+  llvm::SmallVector<RootSignature::ContainerRootParameter, 8>
+      RSParameters;
+  llvm::SmallVector<RootSignature::StaticSamplerDesc, 8> StaticSamplers;
+  llvm::SmallVector<RootSignature::ContainerDescriptorRange, 8>
+      DescriptorRanges;
+  llvm::SmallVector<
+      std::variant<RootSignature::ContainerRootConstants,
+                   RootSignature::ContainerRootDescriptor,
+                   RootSignature::ContainerRootDescriptorTable>,
+      8>
+      RSParamExtras;
+};
```

Comment:
I don't understand how this maps to root signatures.  For example, how would these structs represent this root parameter:

```
DescriptorTable(
    CBV(b1), 
    SRV(t1, numDescriptors = 8, flags = DESCRIPTORS_VOLATILE), 
    UAV(u1, numDescriptors = unbounded, flags = DESCRIPTORS_VOLATILE)),
```

ContainerRootDescriptorTable only has a NumDescriptorRanges field.

Reply:
Added comment for DescriptorRanges.

Reply:
Ah, I see.  I don't understand why this design was chosen rather than a more direct one - eg where ContainerRootDescriptorTable might look like:

```c++
struct ContainerRootDescriptorTable {
  SmallVector<ContainerDescriptorRange> Ranges;
};
```

Reply:
It was started from final serialization format.
Updated to save Ranges in ContainerRootDescriptorTable .